### PR TITLE
Dapr/wip/ch1 world dist

### PR DESCRIPTION
**Pull request by robvet**
## Summary

Please review this PR for Chapter #1 for Dapr for .NET Developers.

Looking forward to your feedback.

### CODE REVIEW

File: docs/architecture/dapr-for-net-developers/the-world-is-distributed.md [hunk 1]
```diff
@@ -2,74 +2,82 @@
...
+In many ways, monolithic apps are `straightforward`. They're straightforward to...
 
-- Coupling
-- Deployment
-- Scaling
+ - build
+ - test
+ - deploy
+ - troubleshoot
+ - scale
```
Comment (amolenk) [inline docs/architecture/dapr-for-net-developers/the-world-is-distributed.md:35]: Up to a point. Most are easy to scale vertically, and can be horizontally scaled somewhat by deploying multiple instances (expensive). But more fine-grained horizontal scaling is not possible.

File: docs/architecture/dapr-for-net-developers/the-world-is-distributed.md [hunk 2]
```diff
@@ -2,74 +2,82 @@
...
+Note how the same application is decomposed across a set of distributed services. Each is self-contained and encapsulates its own code, data, and dependencies. Each is deployed in a software container and managed by a container orchestrator. Instead of a shared database, each service owns it own datastore. Note how some services require a full relational database, but others, a NoSQL datastore. The Basket service stores its state in a distributed key-value cache. Note how inbound traffic routes through an API Gateway service. It's responsible for directing calls to back-end services and enforcing cross-cutting concerns. Most importantly, the application takes full advantage of the scalability, availability, and resiliency features found in modern cloud platforms.
 
 But, while distributed applications can help bring agility and speed, they bring many challenges.
 
 | Challenge | Description |
```
Comment (amolenk) [inline docs/architecture/dapr-for-net-developers/the-world-is-distributed.md:65]: Not sure if we want the challenges to map so perfectly to the Dapr feature set. It feels a bit like we're cheating here ;-)
Also, I don't think things like 'secrets' or 'external resource bindings' are something that one would traditionally identify as a challenge of a distributed system.

Perhaps we can generalize the challenges a bit. For example, 'how to be resilient to failure', 'how to monitor/observe the application as a whole', 'how do service communicate', 'how to ensure scalability'. 
Then when introducing Dapr, we can talk about how the building blocks help solve these challenges.

**Review by amolenk — COMMENTED** [review]
Very nice, I really like the tone of the text and it reads really well.

File: docs/architecture/dapr-for-net-developers/the-world-is-distributed.md (hunk 1 above)
Comment (robvet) [reply docs/architecture/dapr-for-net-developers/the-world-is-distributed.md]: Excellent point. Was thinking along the lines of scaling out a web app with a load balancer in front. How about, 'scale-up vertically'

File: docs/architecture/dapr-for-net-developers/the-world-is-distributed.md (hunk 2 above)
Comment (robvet) [reply docs/architecture/dapr-for-net-developers/the-world-is-distributed.md:65]: Excellent. Refactored list. Tired to generize it. Let me know you thoughts.

Comment (robvet) [reply docs/architecture/dapr-for-net-developers/the-world-is-distributed.md:65]: btw, like the edits too

Comment (amolenk) [reply docs/architecture/dapr-for-net-developers/the-world-is-distributed.md]: I like it much better this way. One suggestion I do have is to combine the 'direct communication' and 'discovery' challenges into one:
" - How can distributed services discover each other and communicate synchronously?"

File: docs/architecture/dapr-for-net-developers/the-world-is-distributed.md [hunk 3]
```diff
@@ -2,74 +2,82 @@
...
+But, this evolution raises many questions...
 
 - What exactly is a distributed application?
-- Why is it important?
+- Why are they gaining popularity?
 - What are the costs?
 - And, importantly, what are the tradeoffs?
 
-For years, we designed, developed, and deployed applications as a single, monolithic unit. Figure 1.x shows a monolithic architecture.
+To start, let's rewind and look at the past 15 years. During this period, we typically constructed applications as a single, monolithic unit. Figure 1-1 shows the architecture.
```
[the remark this answers is not in the document]
Comment (robvet) [reply docs/architecture/dapr-for-net-developers/the-world-is-distributed.md:21]: Great Suggestion. Have added section on SOA. Please review and word-smith on it. Thinking you have more experience with it. Thanks!

File: docs/architecture/dapr-for-net-developers/the-world-is-distributed.md [hunk 4]
```diff
@@ -2,74 +2,82 @@
...
+Note how the same application is decomposed across a set of distributed services. Each is self-contained and encapsulates its own code, data, and dependencies. Each is deployed in a software container and managed by a container orchestrator. Instead of a shared database, each service owns it own datastore. Note how some services require a full relational database, but others, a NoSQL datastore. The Basket service stores its state in a distributed key-value cache. Note how inbound traffic routes through an API Gateway service. It's responsible for directing calls to back-end services and enforcing cross-cutting concerns. Most importantly, the application takes full advantage of the scalability, availability, and resiliency features found in modern cloud platforms.
```
[the remark this answers is not in the document]
Comment (robvet) [reply docs/architecture/dapr-for-net-developers/the-world-is-distributed.md:84]: Really good catch!  Fixed

File: docs/architecture/dapr-for-net-developers/the-world-is-distributed.md [hunk 5]
```diff
@@ -2,74 +2,82 @@
...
 
 | Challenge | Description |
 | :-------- | :-------- |
-| State Management | Maintaining contextual information during a transaction |
-| Interservice Communication | Communicate with other independent services |
-| Messaging |  |
-| Obseraviblity | End-to-end monitoring of processes executing on different machines |
+| Cross-service communication | How can services synchronously communicate with each other using platform agnostic protocols and well-known endpoints? |
+| State Management | How can services maintain contextual information across a transaction? |
+| Asynchronous messaging | How can services implement secure, scalable pub/sub messaging across different kinds of message brokers? |
```
[the remark this answers is not in the document]
Comment (robvet) [reply docs/architecture/dapr-for-net-developers/the-world-is-distributed.md]: Again, great point. Have added verbiage. Please review.

File: docs/architecture/dapr-for-net-developers/the-world-is-distributed.md (hunk 2 above)
Comment (robvet) [reply docs/architecture/dapr-for-net-developers/the-world-is-distributed.md]: Great suggestion. Done. Reads much better.

File: docs/architecture/dapr-for-net-developers/the-world-is-distributed.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (EdwinVW) [reply docs/architecture/dapr-for-net-developers/the-world-is-distributed.md]: I like your changes. Flows better.
But the typo is still in there: "_Instead of a shared database, each service owns --> it**s** <-- own datastore._" 
Suggestion: "_owns its own_" is a bit weird. We could change it to "_Instead of a single database shared by multiple services, each service owns a private database. Other services cannot access this database directly and can only get to data that is exposed through the public API of the service that owns it_".

File: docs/architecture/dapr-for-net-developers/the-world-is-distributed.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (EdwinVW) [reply docs/architecture/dapr-for-net-developers/the-world-is-distributed.md:21]: I think what you added is fine. It stresses the point of the ideas of SOA being valid, but the execution often sub-optimal ("hidden monoliths"). And the emphasis on complex ESBs that contained business-logic is important to mention.

[the remark this answers is not in the document]
Comment (EdwinVW) [reply docs/architecture/dapr-for-net-developers/the-world-is-distributed.md:21]: In the second sentence in this part, I guess a word is missing:
"_In this model, service consumers and service --> **providers** <-- collaborated via middleware messaging components_". 
Maybe only "services" suffices here: "_In this model, services collaborated via middleware messaging components_".

File: docs/architecture/dapr-for-net-developers/the-world-is-distributed.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (EdwinVW) [reply docs/architecture/dapr-for-net-developers/the-world-is-distributed.md]: Great update.
In the final sentence of this part, a word is missing: "_But, shielding your application --> **from** <-- product differences becomes a challenge to keeping code maintainable and portable._".

**Comment by robvet** [comment]
Thinking maybe, do final review of this chapter, and then PR it to Nish for his review. Thoughts?  Working on Ch 2. About 75% complete.

File: docs/architecture/dapr-for-net-developers/the-world-is-distributed.md [hunk 6]
```diff
@@ -2,74 +2,92 @@
...
 description: Gain an understanding of the challenges of distributed applications
 author: robvet
-ms.date: 09/17/2020
+ms.date: 10/18/2020
 ---
 
 # The world is distributed
 
-Modern, distributed systems are in and, monolithic apps are out! Just ask any 'cool kid.'
+Just ask any 'cool kid': *Modern, distributed systems are in and, monolithic apps are out!*
```
Comment (amolenk) [inline docs/architecture/dapr-for-net-developers/the-world-is-distributed.md:10]:
```suggestion
Just ask any 'cool kid': *Modern, distributed systems are in, and monolithic apps are out!* 
```

File: docs/architecture/dapr-for-net-developers/the-world-is-distributed.md [hunk 7]
```diff
@@ -2,74 +2,92 @@
...
-For years, we designed, developed, and deployed applications as a single, monolithic unit. Figure 1.x shows a monolithic architecture.
+To start, let's rewind and look at the past 15 years. During this period, we typically constructed applications as a single, monolithic unit. Figure 1-1 shows the architecture.
 
 ![Monolithic architecture.](./media/monolithic-design.png)
 
-**Figure 1**. Monolithic architecture.
+**Figure 1-1. Monolithic architecture.
 
-In the previous figure, note how modules for Ordering, Identity, Marketing, and others, all reside in a single-server process. Application state is stored inside a shared relational database. Business functionality is exposed via HTML and RESTFul interfaces.
+Note how the modules for Ordering, Identity, and Marketing execute in a single-server process. Application data is stored in a shared database. Business functionality is exposed via HTML and RESTFul interfaces.
```
Comment (amolenk) [inline docs/architecture/dapr-for-net-developers/the-world-is-distributed.md:27]:
```suggestion
Note how the modules for Ordering, Identity, and Marketing execute in a single-server process. Application data is stored in a shared database. Business functionality is exposed via HTML and RESTful interfaces.
```

File: docs/architecture/dapr-for-net-developers/the-world-is-distributed.md [hunk 8]
```diff
@@ -2,74 +2,92 @@
...
+
+![SOA.](./media/soa-basic.png)
+
+**Figure 1-2. SOA architecture.
+
+With SOA, centralized service providers registered with the ESB. Business logic would be built into the ESB to integrate providers and consumers. Service consumers could then find and communicate with these providers using the ESB. 
+
+Despite the promises of SOA, implementing this approach often increased complexity and introduced bottlenecks. Maintenance costs became high and ESB middleware expensive. Services tended to be large. They often shared dependencies and data storage. In the end, SOAs often resulted in a 'distributed monolithic' structure with centralized services that were resistant to change. 
+
+Nowadays, many organizations have realized speed and agility by adopting a distributed microservice architectural approach to building systems. Figure 1-3 shows the same system built applying distributed techniques and practices.
```
Comment (amolenk) [inline docs/architecture/dapr-for-net-developers/the-world-is-distributed.md:65]:
```suggestion
Nowadays, many organizations have realized speed and agility by adopting a distributed microservice architectural approach to building systems. Figure 1-3 shows the same system built using distributed techniques and practices.
```

**Review by amolenk — COMMENTED** [review]
Just a couple of small things. Nice introduction chapter, let's ship it!

File: docs/architecture/dapr-for-net-developers/the-world-is-distributed.md (hunk 6 above)
Comment (robvet) [reply docs/architecture/dapr-for-net-developers/the-world-is-distributed.md:10]: Great catch - corrected

File: docs/architecture/dapr-for-net-developers/the-world-is-distributed.md (hunk 7 above)
Comment (robvet) [reply docs/architecture/dapr-for-net-developers/the-world-is-distributed.md:27]: Another great catch. Corrected.

File: docs/architecture/dapr-for-net-developers/the-world-is-distributed.md (hunk 8 above)
Comment (robvet) [reply docs/architecture/dapr-for-net-developers/the-world-is-distributed.md:65]: Great. Updated

**Pull request merged**